File: pkg/algorithms/quicksort/quicksort.go
package quicksort

func sort(list []int, begin int, end int) {
	pivot := list[(begin+end)/2]

	left := begin
	right := end
	for left < right {
		if list[left] < pivot {
			left += 1
		}

		if list[right] > pivot {
			right -= 1
		}

		if left < right {
			aux := list[left]
			list[left] = list[right]
			list[right] = aux
		}
	}

	if left < end {
		sort(list, left+1, end)
	}

	if right > begin {
		sort(list, begin, right)
	}
}

func Sort(list []int) {
	sort(list, 0, len(list)-1)
}

File: algorithms/quicksort/src/quicksort.cpp
#include <stdio.h>

#include "./quicksort.h"

int Quicksort::partition(int *vector, int left, int right)
{
    int pivot = vector[right];
    int pointer = left;

    for (int index = left; index < right; index++)
    {
        if (vector[index] < pivot)
        {
            int aux = vector[pointer];
            vector[pointer] = vector[index];
            vector[index] = aux;

            pointer += 1;
        }
    }

    int aux = vector[pointer];
    vector[pointer] = vector[right];
    vector[right] = aux;

    return pointer;
}

void Quicksort::sort(int *vector, int left, int right)
{
    if (left < right)
    {
        int pivot = Quicksort::partition(vector, left, right);

        Quicksort::sort(vector, pivot + 1, right);
        Quicksort::sort(vector, left, pivot - 1);
    }
}